### ISSUE TITLE
Gradual memory leak with no plateau

### ISSUE THREAD

**Issue by davidcelis** [octobox/octobox#667]
Are you experiencing this issue in octobox.io or your own instance?

* ~octobox.io~
* My instance

🐞 **Problem**

I run a personal Octobox instance on a DigitalOcean droplet (4GB of RAM) so that I can take advantage of the `FETCH_SUBJECTS` feature. On this instance, I also have personal access tokens enabled and am using one that I generated specifically for Octobox. My method for running it is via `docker-compose` (I configured all of the environment variables in a `docker-compose.yml` file).

After running the instance for a week or so, I started getting errors during syncing. The error was that the `pg` gem couldn't open connections to postgres due to a lack of RAM. Unfortunately, the server's RAM was so tied up that I couldn't even open an SSH connection; I had to log into DigitalOcean and perform a hard reboot 😰

After the hard reboot, the server and Octobox were up and running again. However, RAM continues to climb. This is a graph of memory usage since I rebooted the instance on the 20th:

![image](https://user-images.githubusercontent.com/36873/41869970-87438614-786f-11e8-9a77-93845fd74a19.png)

My assumption is that you haven't noticed a memory leak on octobox.io so, if that's the case, I also assume that it has to do with the `FETCH_SUBJECTS` feature.

🎯 **Goal**

Octobox, even with the experimental `FETCH_SUBJECTS` feature enabled, does not leak memory

💡 **Possible solutions**

A separate instance with memory debugging tools installed should probably be monitored for a short time to deduce the source of the leak

📋  **Steps to solve the problem**

Monitor an Octobox instance for memory leaks and figure out what resources are hanging around unnecessarily

**Comment by chrisarcand** [comment]
> My assumption is that you haven't noticed a memory leak on octobox.io so, if that's the case, I also assume that it has to do with the FETCH_SUBJECTS feature.

Given the sheer amount of functionality that sits in that feature, probably - but more than that, I'm guessing that octobox.io and others don't have this problem because Heroku - IIRC dynos are restarted every 24 hours whether you want it or not.

**Comment by davidcelis** [comment]
> IIRC dynos are restarted every 24 hours whether you want it or not.

Ohhhh yup, you're right https://devcenter.heroku.com/articles/dynos#restarting

**Comment by andrew** [comment]
One other thing I've done to reduce memory usage on octobox.io is to switch to using jemalloc, might be worth trying that out before the more crude cronjob to restart it every 24 hours.

Used this build pack on heroku: https://github.com/gaffneyc/heroku-buildpack-jemalloc

**Comment by Alexkho82** [comment]
#667

**Comment by pradyunsg** [comment]
I just had my Heroku instance with ~150 unread notifications kill a worker due to running out of memory -- not sure if it's related to this directly.

```
2018-09-11T13:05:03.428549+00:00 app[web.1]: [4] PumaWorkerKiller: Out of memory. 2 workers consuming total: 508.3671875 mb out of max: 501.76 mb. Sending TERM to pid 15 consuming 210.05078125 mb.
```

Another thing that happened simultaneously - a timed-out request:

```
2018-09-11T13:05:13.396753+00:00 heroku[router]: sock=client at=warning code=H27 desc="Client Request Interrupted" method=GET path="/" host=octobox.pradyunsg.me request_id=562a06ad-53d7-42f1-a747-cd197becc0fe fwd="***.***.***.***" dyno=web.1 connect=1ms service=1164ms status=499 bytes= protocol=http
2018-09-11T13:05:26.893441+00:00 heroku[router]: at=error code=H12 desc="Request timeout" method=GET path= "/notifications/sync?async=false" host=octobox.pradyunsg.me request_id=52fe03b0-d620-4e27-aa71-5f941e55e180 fwd="***.***.***.***" dyno=web.1 connect=1ms service=30004ms status=503 bytes=0 protocol=http
[2 lines...]
2018-09-11T13:05:33.883526+00:00 app[web.1]: [2018-09-11 13:05:33 +0000] INFO -- pradyunsg synced their notifications
```

I've restarted all dynos and now the usage seems to be much lower. There is still a lot of latency involved though:

```
2018-09-11T13:15:03.737698+00:00 heroku[router]: at=info method=GET path="/" host=octobox.pradyunsg.me request_id=a20d531e-fd5e-4b4e-944b-5bb71e33aace fwd="118.185.217.195" dyno=web.1 connect=1ms service=582ms status=200 bytes=95908 protocol=http
```

---

@andrew Are these related to this issue or should I file a new one?

**Comment by andrew** [comment]
@pradyunsg there might be some differences between this original issue and the one your hitting on heroku, also not sure if this issue is still happening to @davidcelis?

On octobox.io I'm running this setup with professional dynos:

![screen shot 2018-09-11 at 14 25 49](https://user-images.githubusercontent.com/1060/45363144-9e635300-b5ce-11e8-9ae2-24484e6254f7.png)

With `ENV['RAILS_MAX_THREADS']` set to 6

I'd guess that your worker process ran out of ram, can you share the last two hours metrics graph for the worker? 

Reducing RAILS_MAX_THREADS to 3 will help reduce ram and shouldn't impact anything if you're the only person using that instance.

**Comment by pradyunsg** [comment]
> I'd guess that your worker process ran out of ram, can you share the last two hours metrics graph for the worker?

I'm using on the free dynos, so I don't have any metrics. :(

OTOH, free dynos have 512 MB RAM and the log message shows 501.76 MB being consumed, so I'm pretty certain that is the case.

> Reducing RAILS_MAX_THREADS to 3 will help reduce ram and shouldn't impact anything if you're the only person using that instance.

ISTM that the notification syncing logic is not happy anyway since it is still timing out. I'll try this suggestion, hopefully it works. I'll open a new issue otherwise.

**Comment by andrew** [comment]
FYI [puma_worker_killer](https://github.com/schneems/puma_worker_killer#configure) is enabled by default when heroku is deployed to production, and the threshold is at 98%, 512*0.98 = 501.76 so that makes sense.

**Comment by pradyunsg** [comment]
FTR, I set notification sync to be done in the background using sidekiq (OCTOBOX_BACKGROUND_JOBS_ENABLED) and now it seems to be all good. :)

**Comment by davidcelis** [comment]
> not sure if this issue is still happening to @davidcelis?

It is; I still have a cron job rebooting the app once per day to avoid it. I have not tried setting `OCTOBOX_BACKGROUND_JOBS_ENABLED` though!

**Comment by chrisarcand** [comment]
I'll be moving off Heroku to a setup similar to @davidcelis, so more eyes on this soon, hopefully.

**Comment by andrew** [comment]
@davidcelis I wonder if your docker image is running in development mode, that would definitely hamper things.

**Comment by andrew** [comment]
I also thought I'd have some fun with https://github.com/SamSaffron/memory_profiler this evening to see if I could spot anything, partly because on the octobox.io heroku, the sidekiq workers need twice as much ram (1GB) than the rails web process (~300mb) with `RAILS_MAX_THREADS` set to 6.

One thing I did notice is that [sawyer](https://github.com/lostisland/sawyer), the only dependency of octokit, allocates a lot of objects when syncing notifications, specifically in the [link_parser](https://github.com/lostisland/sawyer/blob/master/lib/sawyer/link_parsers/simple.rb) which we don't make any use of. 

So I thought I'd see if I could swap out the Octokit call in [download_service](https://github.com/octobox/octobox/blob/master/app/services/download_service.rb#L39) with something simpler:

```ruby
params[:access_token] = client.access_token
params[:per_page] = '100'
query = URI.encode_www_form(params)
JSON.parse(Typhoeus.get('https://api.github.com/notifications?' + query).body)
```

Obviously that throws out a number of features we do actually use, like pagination and error handling, but the memory usage is significantly different in a simple benchmark, syncing, then deleting my notifications (60 or so) three times:

```
with Octokit

Total allocated: 24403328 bytes (243821 objects)
Total retained:  1280334 bytes (2527 objects)

without Octokit

Total allocated: 8828774 bytes (97076 objects)
Total retained:  1273584 bytes (2362 objects)
```

That's about 3x less total allocated, so might be worth investing some time in a cut down version of octokit that skips sawyer, given that Octobox spends a lot of downloading things from the GitHub API.

**Comment by davidcelis** [comment]
@andrew The docker image is definitely running in development mode; it's hardcoded to do so (the image does not install production gems), and the deployment instructions also use development as the value for `RAILS_ENV` 😅

**Comment by chrisarcand** [comment]
☝️ To be fixed in #927
